### PR TITLE
Move the syntax modules into a separate library

### PR DESCRIPTION
**Pull request by MartyO256**
This PR is the second step in refactoring out the parser for the Beluga/Harpoon external syntax.
It is a continuation of #245.

The primary goal of this PR is to move the various modules relating to syntaxes for Beluga and Harpoon into a separate library.
Specifically, the modules `gensym`, `holeId`, `id`, `synapx`, `syncom`, `synint` and `syntax` are moved from `./src/core` to `./src/syntax`.
This will allow for the external syntax parser modules to be refactored into a separate `parser` library that won't depend on the core modules.

The `Location` module was being re-exported in multiple places from the syntax modules.
Import statements for that module were normalized to avoid confusion.

### CODE REVIEW

**Comment by MartyO256** [comment]
There are two cases of non-exhaustive pattern-matching in the modules that were moved to the `syntax` library.
They both occur in module `synint`, and have to do with LF and COMP spines.
For now, these functions were explicitly annotated with `[@warning "-8"]`.

https://github.com/MartyO256/Beluga/blob/3615adc423c95abbf032e0f9f5f2a139c2b94699/src/syntax/synint.ml#L370-L374

https://github.com/MartyO256/Beluga/blob/3615adc423c95abbf032e0f9f5f2a139c2b94699/src/syntax/synint.ml#L603-L607

File: src/syntax/syntax.ml [hunk 1]
```diff
@@ -3,10 +3,11 @@
...
-module Loc = Location
-
 module Common = Syncom
 module Int = Synint
 module Apx = Synapx
 module Ext = Synext
+
+module Id = Id
+module HoleId = HoleId
+module Gensym = Gensym
```
Comment (Ailrun) [inline src/syntax/syntax.ml:13]: The module name (`Syntax.Syntax. ...`) is a bit confusing... I think giving some different name to either the library or module would be easier to understand.

**Comment by Ailrun** [comment]
I'm afraid that if this breaks what @sjjs7 work on though...

**Comment by MartyO256** [comment]
This would break the `aps` branch @sjjs7 is working on because the `Synext` module would be referred to using `Syntax.Ext` and because `MQuery` has to be added to the external syntax.
A git rebase would be required.

File: src/syntax/syntax.ml (hunk 1 above)
Comment (Ailrun) [reply src/syntax/syntax.ml:13]: I mean `(wrapped true)` itself is fine; or rather, it is preferable. What I want to mention is just its naming. Restructure entire `syntax` library modules would also be a solution, but I think renaming either the library (to something like `SyntaxLib`) or the main module (to something like `SyntaxMain`, `SyntaxTop`, `Syn`, or any other different names) might be easier to be brought into this PR.

**Pull request closed without merging**